geode-core/src/main/java/org/apache/geode/internal/cache/TXCommitMessage.java

```diff
@@ -813,16 +812,17 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
     this.farsideBaseThreadId = in.readLong();
     this.farsideBaseSequenceId = in.readLong();
 
-
     this.needsLargeModCount = in.readBoolean();
 
+    boolean hasShadowKeys = in.readBoolean();
```

Comment:
Don't you need a version check here & not read the shadowkeys boolean if the stream version is < 1.8?

Reply:
I think you are correct. During a rolling upgrade an old peer could send a new peer a TXCommitMessage that does not have the shadowKeys boolean

Reply:
which means we would need to keep a way of determining if we have shadow keys when receiving a TXCommitMessage from an older member. Since clients never send TXCommitMessage would could just set hasShadowKeys to true if the inputStream says it is coming from a member older than 1.8

Reply:
You're right. We aren't covering the case where an old server is sending to a new server. We do have the other way around though. Should be an easy fix.

Reply:
✔️